nixos/modules/system/boot/networkd.nix

```diff
@@ -3181,6 +3181,29 @@ let
       ];
       kernelModules = [ "af_packet" ];
 
+      systemd.services.nixos-flush-networkd = mkIf config.boot.initrd.network.flushBeforeStage2 {
+        description = "Flush Network Configuration";
+        wantedBy = ["initrd.target"];
+        after = ["systemd-networkd.service" "dbus.socket" "dbus.service"];
+        before = ["shutdown.target" "initrd-switch-root.target"];
+        conflicts = ["shutdown.target" "initrd-switch-root.target"];
+        unitConfig.DefaultDependencies = false;
+        serviceConfig = {
+          # This service does nothing when starting, but brings down
+          # interfaces when switching root. This is the easiest way to
+          # ensure proper ordering while stopping
+          Type = "oneshot";
```

Comment:
Can you link to the docs describing this should be flushed like this? I'm also surprised we need to manually parse the output and networkctl down individually - is there no command to do this better? Also, what about network mounts set up during initrd?

Reply:
`networkctl` does not have any command / option for bringing down all interfaces, unfortunately.

Reply:
Updated comment with reference to systemd documentation on the matter.

@flokli As for network mount units, that's already broken if you use `flushBeforeStage2` with scripted initrd. With systemd initrd, `flushBeforeStage2` is disabled by default though. And yes, I should think you would expect a network file system to break if you enable the option that brings down networking :P

Reply:
I meant docs on the "flushing networkd" part, and that there's no command for it ;-)

Reply:
ah

Reply:
Oh, then I'm not sure what you want linked? The `flushBeforeStage2` option means "I want interfaces brought down before stage 2", so `networkctl down` is how you do that. Not really anything to link to. I can leave a comment mentioning that there's no `--all` argument or anything like that though.

Reply:
> Oh, then I'm not sure what you want linked?

Documentation about why we need to do the manual flushing in first place ;-)

I looked at https://github.com/systemd/systemd/issues/780#issuecomment-166148078, which suggests flushing has been implemented in networkd, and there's even possibilities to have it not touch certain interfaces (`CriticalInterface`)

Reply:
> I looked at [systemd/systemd#780 (comment)](https://github.com/systemd/systemd/issues/780#issuecomment-166148078), which suggests flushing has been implemented in networkd, and there's even possibilities to have it not touch certain interfaces (`CriticalInterface`)

@flokli reads to me like the discussion over there is about deconfiguring addresses when networkd *first takes possession* of an interface, not about any behaviour wrt. to shutdown.  
They don't seem to use the word “flush” in the same sense as it is used here.

Reply:
Comment added. @flokli Does that answer the question?

Reply:
Sorry, i still dont understand why it's desirable to flush, and why this is on by default.

If i use network in initrd, I'd expect systemd-networkd in initrd config to be semi-compatible with the one on my real system, so i don't quite understand what we gain from flushing all addresses, and this is not very obvious from reading the option description. What's the actual breakage for a "default user" (probably doing DHCP on all wired interfaces) if this is left on?

Also note this only brings interfaces down (and maybe removes addresses or links)? It doesn't delete stacked interfaces like vlans or bonds.

Reply:
> Sorry, i still dont understand why it's desirable to flush, and why this is on by default.

@flokli Well, for one thing, this PR changes it so it *isn't* on by default if systemd initrd is enabled. And whether or not the desire to flush makes sense is sort of immaterial to this PR, no? The option has existed in the scripted initrd for a long time. This PR merely implements the feature, and disables it by default with systemd initrd because you're right, it doesn't really make a whole lot of sense most of the time I don't think

Reply:
> Sorry, i still dont understand why it's desirable to flush, and why this is on by default.

Because the stage2 may want to do things with the interfaces, that it cannot do, while they're still up, like for example renaming them or setting other netdev properties.

Reply:
I'd have assumed renaming happens by `.link` units (parsed by udev), way before stage2 needs to do it. See also 355c58e485e76a4b91bb4613ee2a3e1bae809d28.

What's it that you can't express with `.link` files? Maybe it could become part of the description, so it's more understandable where this is necessary.

In any case, as this flushing is disabled by default, and opt-in, it's fine now IMHO.